src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs

```diff
@@ -45,36 +51,44 @@ namespace Microsoft.VisualStudio.LanguageServices.Packaging
     [ExportWorkspaceService(typeof(IPackageInstallerService)), Shared]
     internal partial class PackageInstallerService : AbstractDelayStartedService, IPackageInstallerService, IVsSearchProviderCallback
     {
-        private readonly object _gate = new object();
         private readonly VisualStudioWorkspaceImpl _workspace;
         private readonly SVsServiceProvider _serviceProvider;
+        private readonly Shell.IAsyncServiceProvider _asyncServiceProvider;
         private readonly IVsEditorAdaptersFactoryService _editorAdaptersFactoryService;
 
         private readonly Lazy<IVsPackageInstallerServices> _packageInstallerServices;
         private readonly Lazy<IVsPackageInstaller2> _packageInstaller;
         private readonly Lazy<IVsPackageUninstaller> _packageUninstaller;
         private readonly Lazy<IVsPackageSourceProvider> _packageSourceProvider;
 
-        private JoinableTask<ImmutableArray<PackageSource>?> _packageSourcesAsync;
-        private IVsPackage _nugetPackageManager;
-
-        private CancellationTokenSource _tokenSource = new CancellationTokenSource();
+        private IVsPackage? _nugetPackageManager;
 
         // We keep track of what types of changes we've seen so we can then determine what to
         // refresh on the UI thread.  If we hear about project changes, we only refresh that
         // project.  If we hear about a solution level change, we'll refresh all projects.
-        private bool _solutionChanged;
-        private readonly HashSet<ProjectId> _changedProjects = new HashSet<ProjectId>();
+        private AsyncBatchingWorkQueue<(bool solutionChanged, ProjectId? changedProject)>? _workQueue;
 
         private readonly ConcurrentDictionary<ProjectId, ProjectState> _projectToInstalledPackageAndVersion =
             new ConcurrentDictionary<ProjectId, ProjectState>();
 
+        /// <summary>
+        /// Lock used to protect reads and writes of <see cref="_packageSourcesTask"/>.
+        /// </summary>
+        private readonly object _gate = new object();
+
+        /// <summary>
+        /// Task uses to compute the set of package sources on demand when asked the first time.  The value will be
+        /// computed and cached in the task.  When this value changes, the task will simply be cleared out.
+        /// </summary>
+        private Task<ImmutableArray<PackageSource>>? _packageSourcesTask;
+
         [ImportingConstructor]
         [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
         public PackageInstallerService(
             IThreadingContext threadingContext,
             VisualStudioWorkspaceImpl workspace,
             SVsServiceProvider serviceProvider,
+            [Import("Microsoft.VisualStudio.Shell.Interop.SAsyncServiceProvider")] object asyncServiceProvider,
             IVsEditorAdaptersFactoryService editorAdaptersFactoryService,
             [Import(AllowDefault = true)] Lazy<IVsPackageInstallerServices> packageInstallerServices,
```

Comment:
AllowDefault means we can potentially get a null here, so this now should be nullable annotated. I don't know off the top of my head if that means the Lazy itself is null, or the value of the lazy is null. :-/ The lack of null checking in other parts of the file makes me believe this was always broken. :frowning:

I'm also not sure if it's critical to have the AllowDefaults anymore: that might be implying that NuGet is an optional part of the VS install that we need to gracefully handle, but this code didn't handle it gracefully. It's also possible it at this point is essentially a dependency of us that cant be uninstalled...

Reply:
Not going to touch that here in this PR :)

Reply:
Agreed. The nullable annotation is incorrect, and it's definitely in scope of the pull request because this pull request is where `#nullable enable` was added. This annotation needs to be fixed before merge.

Reply:
done.